### ISSUE TITLE
Many NaN Losses During Training

### ISSUE THREAD

**Issue by Everloom-129** [priyasundaresan/sphinx#7]
### Description
When training the SPHINX model using the provided code, I'm encountering numerous NaN losses during the training process. This appears to be happening in both waypoint and dense training modes.

for example:
```bash
########################################################
8: click/acc_top20         [  221]: avg:   0.0062, min:   0.0016[ 204], max:   0.0125[ 157]748809823456034029568.0000[  -1]
8: mode/loss               [  221]: avg:      nan, min: 99999999999999997748809823456034029568.0000[  -1], max: -99999999999999997748809823456034029568.0000[  -1]
8: rot_pos/num_postive     [  221]: avg:   6.8199, min:   6.3147[  14], max:   7.3013[  91]
8: rot_pos/pos_loss        [  221]: avg:      nan, min: 99999999999999997748809823456034029568.0000[  -1], max: -99999999999999997748809823456034029568.0000[  -1]
8: rot_pos/rot_loss        [  221]: avg:      nan, min: 99999999999999997748809823456034029568.0000[  -1], max: -99999999999999997748809823456034029568.0000[  -1]
8: train/ema_decay         [  221]: avg:   0.9961, min:   0.9959[   1], max:   0.9963[ 221]
|----------+-------+---------------+-------|
| model    |   221 |         774.6 |  60.5 |
########################################################
9: click/acc_top20         [  221]: avg:   0.0064, min:   0.0016[  10], max:   0.0109[ 140]748809823456034029568.0000[  -1]
9: mode/loss               [  221]: avg:      nan, min: 99999999999999997748809823456034029568.0000[  -1], max: -99999999999999997748809823456034029568.0000[  -1]
9: rot_pos/num_postive     [  221]: avg:   6.8339, min:   6.3391[ 112], max:   7.2670[ 195]
9: rot_pos/pos_loss        [  221]: avg:      nan, min: 99999999999999997748809823456034029568.0000[  -1], max: -99999999999999997748809823456034029568.0000[  -1]
9: rot_pos/rot_loss        [  221]: avg:      nan, min: 99999999999999997748809823456034029568.0000[  -1], max: -99999999999999997748809823456034029568.0000[  -1]
9: train/ema_decay         [  221]: avg:   0.9965, min:   0.9963[   1], max:   0.9966[ 221]
9: train/grad_norm         [  221]: avg:      nan, min: 99999999999999997748809823456034029568.0000[  -1], max: -99999999999999997748809823456034029568.0000[  -1]
Timer Info:
| name     |   num |   t/call (ms) |     % |
|----------+-------+---------------+-------|
| data     |   221 |           2.5 |   0.2 |
| fps      |   221 |         504   |  39.3 |
| model    |   221 |         774.7 |  60.4 |
| stat     |   221 |           0.6 |   0   |
| total(s) |     1 |         283.3 | 100   |
########################################################


```

The NaN losses seems:
1. Click loss calculation
2. Points offset loss calculation
3. Position and rotation losses

Based on `scripts/train_waypoint.py`, I think the NaNs might be due to:
1. Division by zero when normalizing user clicks (`user_clicked_labels / user_clicked_labels.sum(dim=1, keepdim=True)`)
2. Division by zero in masked point offset loss calculation
3. Potential numerical instability in the transformer model

### Questions
1. Is this a known issue with the training process?
2. Are there any recommended hyperparameter settings to prevent these NaNs?
3. Should there be additional numerical stability checks in the loss calculations?

### GPT Suggested Fix
Would it make sense to add numerical stability safeguards such as:
```python
# For click loss
target_user_clicked = user_clicked_labels / (user_clicked_labels.sum(dim=1, keepdim=True) + 1e-8)

# For points offset loss
points_off_loss = (points_off_loss.sum(2) * points_mask).sum(1) / (points_mask.sum(1) + 1e-8)
```

**Comment by hengyuan-hu** [comment]
I think that is likely due to having too few salient points in some of the training examples. 

https://github.com/priyasundaresan/sphinx/blob/43ccae44d138f263709e39af2e0d1e8e36f55f1f/dataset_utils/waypoint_dataset.py#L151
Here we set mark the points within a 0.05 radius (points within a 5cm ball centered around the user-click) as salient point. If there are too few points here, like 50 out of 20K, then it will cause the training to be very unstable or produce NAN because the we would have even fewer or maybe 0 labels after furthest point sampling.

It is generally a bad thing to have too few salient points, so maybe it is helpful to first check the data and see if the salient points (probably marked as red in some of the visualization code) make sense.

The radius is set here. https://github.com/priyasundaresan/sphinx/blob/43ccae44d138f263709e39af2e0d1e8e36f55f1f/dataset_utils/waypoint_dataset.py#L187

**Comment by Everloom-129** [comment]
thanks for the detailed reply! It must be some misunderstanding during our data collection process, thanks we will try again
